### ISSUE TITLE
Docker publish assumes sudo is installed

### ISSUE THREAD

**Issue by exocom** [drone/drone#817]
First off thanks for adding the docker publish plugin. I think it will be really useful for many people.

However the docker publish plugin doesn't seem to be compatible with all base images. When using the default node image the following error is produced:

```
$ type -p docker || wget -qO- https://get.docker.io/builds/Linux/x86_64/docker-1.3.tgz |sudo tar zxf - -C /
/usr/local/bin/drone: line 86: sudo: command not found
```

Here is my .drone.yml

```
image: node:latest
script:
  - echo $DRONE_COMMIT
  - npm install gulp -g
  - npm install
  - gulp test
publish:
   docker:
    dockerfile: Dockerfile
    docker_host: 0.0.0.0
    docker_port: 2375
    docker_version: 1.3
    registry_host: mysite.com
    registry_protocol: https
    registry_login: true
    registry_login_uri: /registry/v1/
    username: $$registry_user
    password: $$registry_pass
    image_name: myImage
    push_latest: true
    keep_builds: false
    tag: $$DRONE_COMMIT
```

**Comment by bradrydzewski** [comment]
I think one issue is that there is no `node:latest` image. You can check here:
https://registry.hub.docker.com/u/bradrydzewski/node/tags/manage/

I wonder which image is actually being used ...

Either way there is a bug. The Docker plugin should have two versions of each command. One that uses `sudo` if it exits, and one that does not. We follow this pattern for many of the other plugins already. For example:

```sh
[ -f /usr/bin/sudo ] && sudo docker ...
[ -f /usr/bin/sudo ] || docker ...
```

**Comment by exocom** [comment]
@bradrydzewski node:latest is library/node:latest === library/node:0.10.35, however, after more experimenting I've discovered the issue.

1. The plugin assumes you have docker CLI or a docker in docker container.
2. The plugin assumes you have the required programs/libraries such as wget, tar, etc.
3. The same issue exists for Drone in general. If you don't have git installed in the base image the build fails.

I think it would be nice if the plugins and drone would check for the dependencies and fail with a nice error message, **before performing any** of the build steps. For example

**Failed to build: "Docker Publish" plugin requires the following programs:**
*docker cli -or- docker running in privileged mode*
*wget*
*tar*
*etc.*

Especially GIT because This was not obvious, but it makes sense after running into it.

**Failed to build because: Hey, we didn't find git installed in the image you provided, your going to need that in order for us to clone your project in here**

I know there are lot of conversations going on in other threads esp about how the build / setup process in general, but it is not clear that plugins require the container images to have certain programs installed in order to be used. In other words you can't just build using the standard library/* images because many of them don't have basic programs like git, make, gcc, etc.

Hope this is helpful

**Comment by bradrydzewski** [comment]
> The plugin assumes you have docker CLI or a docker in docker container.

the plugin will proactively download the docker CLI when not found:
https://github.com/drone/drone/blob/master/plugin/publish/docker.go#L64

the plugin does require you to run the container in privileged mode, which is non-obvious. I haven't used this plugin myself, so I could use some help with the documentation from someone that has experience with the plugin:
http://readme.drone.io/usage/publish/docker/

> I think it would be nice if the plugins and drone would check for the dependencies and fail with a nice error message, before performing any of the build steps. For example

Yes definitely, I think there is a level of polish we are working toward. We're not there yet, but I completely agree with you.

> In other words you can't just build using the standard library/* images because many of them don't have basic programs like git, make, gcc, etc.

If you come across a language stack that doesn't include git, for example, I definitely recommend opening up an issue, and feel free to /cc me. We've gotten both the Java and Go teams to agree to switching their base image to `buildpack-deps:jessie-micro` which includes many essential utilities such as git, wget, and more

**Comment by bradrydzewski** [comment]
fixed in 0.4 release with addition of the [drone-docker](https://github.com/drone-plugins/drone-docker) plugin, which runs docker-in-docker and doesn't require sudo
